### DISCUSSION TITLE
This function self.ble.get_public_key_generated() gives TypeError: function takes 0 positional arguments but 1 were given

### DISCUSSION THREAD

**Discussion by brianreinhold** [micropython/micropython#15945] [RP2040 / Pico]
This is an exercise in accessing a C method from a micropython application

I have this micropython application code 
```
        self.ble: BLE = BLE()
        print(dir(self.ble))
        result: bool = self.ble.get_public_key_generated()
```
which gives this error:  `TypeError: function takes 0 positional arguments but 1 were given`. Clearly no arguments are given.

What the method should do is return true if a set of cryptography keys have been generated by the btstack. The results of the `print(dir(self.ble))` certainly show that the method exists (it's next to last in the list):
```
['__class__', 'active', 'config', 'gap_advertise', 'gap_connect', 'gap_delete_bonding', 'gap_disconnect', 'gap_pair', 'gap_passkey', 'gap_scan', 'gap_set_passkey', 'gattc_discover_characteristics', 'gattc_discover_descriptors', 'gattc_discover_services', 'gattc_exchange_mtu', 'gattc_read', 'gattc_write', 'gatts_indicate', 'gatts_notify', 'gatts_read', 'gatts_register_services', 'gatts_set_buffer', 'gatts_write', 'get_public_key_generated', 'irq']
```
The method is defined in `lib/btstack/src/btstack_crypto.c` and declared in `lib/btstack/src/btstack_crypto.h`. Very simple.
```
uint8_t get_public_key_generated(void)
{
    return btstack_crypto_ecc_p256_public_key_generated;
}
```
To access it in micropython at the application level one has the usual hierarchy
First the mapping of the function from python to C and back which appears in the `modbluetooth.c` file:
```
STATIC mp_obj_t bluetooth_ble_get_public_key_generated(void) 
{
    uint8_t result = mp_bluetooth_get_public_key_generated();
    return mp_obj_new_bool(result);
}

STATIC MP_DEFINE_CONST_FUN_OBJ_0(bluetooth_ble_get_public_key_generated_obj, bluetooth_ble_get_public_key_generated);
```
where the macro `STATIC MP_DEFINE_CONST_FUN_OBJ_0` as I understand it is supposed to handle a method that takes 0 arguments.

Second the inclusion of this method in the `STATIC const mp_rom_map_elem_t bluetooth_ble_locals_dict_table[]` as the following entry:
`{ MP_ROM_QSTR(MP_QSTR_get_public_key_generated), MP_ROM_PTR(&bluetooth_ble_get_public_key_generated_obj) },` 
and its declaration in `modbluetooth.h` 
`uint8_t mp_bluetooth_get_public_key_generated(void);`

And finally it's presence in the btstack-specific modbluetooth file `modebluetooth_btstack.c`
```
uint8_t mp_bluetooth_get_public_key_generated()
{
    uint8_t result = get_public_key_generated();
    DEBUG_printf("mp_bluetooth_get_public_key_generated:%u\n", result);
    return result;
}
```
Now it's certainly easy to get messed up on the `mp_bluetooth_*`, `bluetooth_ble_*` and its `obj` signatures but as far as I see it is all done correctly. This method is clearly setup to take 0 arguments and it compiles and builds with no warnings or error messages. Yet when I call it from a micropython application I get an error saying I am passing in an argument which is clearly not the case.

I have spent hours examining the code, redoing the firmware build always getting the same result. I am clueless. Must be something stupid I do not see because I have looked at it too long. Any insight from someone who is better versed in this python to C interfacing is GREATLY appreciated!

**Comment by andrewleech** [comment]
Hi, I think the issue here is that the function is being run as a method of a class instance - in this case there _is_ av single argument of `self` passed implicitly behind the scenes. 

Try running it as `BLE.get_public_key_generated()` ie as a static class method, rather than an instance method. Not sure if that'll actually work but worth a try. 

Otherwise the C functions can be declared as accepting 1 arg and just ignore the arg.

**Reply by brianreinhold** [reply]
Out of frustration I thought of passing one argument and ignoring the argument (but have not done it) and using the `STATIC MP_DEFINE_CONST_FUN_OBJ_1` macro. If I use `BLE.` wont that require a static signature? I have `bluetooth_ble` all over the place which makes it part of the ble instance. I have defined another method
```
void mp_bluetooth_delete_bonding(uint8_t addr_type, uint8_t *addr) {
    DEBUG_printf("mp_bluetooth_delete_bonding in modbluetooth_btstack.c:%u\n", addr_type);
    gap_delete_bonding(addr_type, addr);
}
```
which takes two arguments going though a similar but more complicated mess because of the parameters and that does work.
I have also defined a method `BLE.gap_set_passkey(999997)` which is static and has to be called with `BLE.` but it has a different set of signatures in the C code.

I just don't understand what is going on.

**Reply by brianreinhold** [reply]
This is the ugliest solution (hack fix in my opinion). I changed
```
STATIC mp_obj_t bluetooth_ble_get_public_key_generated(void) 
{
    uint8_t result = mp_bluetooth_get_public_key_generated();
    return mp_obj_new_bool(result);
}

STATIC MP_DEFINE_CONST_FUN_OBJ_0(bluetooth_ble_get_public_key_generated_obj, bluetooth_ble_get_public_key_generated);
```
to this
```
STATIC mp_obj_t bluetooth_ble_get_public_key_generated(size_t n_args, const mp_obj_t *args) 
{
    uint8_t result = mp_bluetooth_get_public_key_generated();
    return mp_obj_new_bool(result);
}

STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(bluetooth_ble_get_public_key_generated_obj, 0, 1, bluetooth_ble_get_public_key_generated);
```
It handles the mystery 'self' variable assumed being passed to the BLE class (which does not exist).

Now that I got it to work after 12 hours the function itself does not! Too bad it takes 10 minutes per build.

**Comment by bidrohini** [comment]
Try to call the method without the self reference, for example:
result = BLE.get_public_key_generated()
